### PR TITLE
refactor(connector): Use connector_response_reference_id  for Shift4

### LINKED ISSUE #2316: [FEATURE]: [Shift4] Use `connector_request_reference_id` as reference to the connector
### :memo: Feature Description 

- We transmit our reference id from Hyperswitch to the connector, enabling merchants to subsequently locate the payment within the connector dashboard.
- Presently, there is a lack of uniformity in this process, as some implementations employ `payment_id` while others utilize `attempt_id`.
- This inconsistency arises from the fact that `payment_id` and `attempt_id` correspond to two distinct entities within our system.

### :hammer: Possible Implementation 

- To mitigate this confusion, we've introduced a new field called `connector_request_reference_id` in our RouterData. 
- This id should replace them as a reference id to the connector for payment request. 
- One might need to have exposure to api docs of the connector for which it is being implemented.
- You can check this PR for further reference https://github.com/juspay/hyperswitch/pull/2052

:bookmark: Note : All the changes needed should be contained within `hyperswitch/crates/router/src/connector/`

### :package: Have you spent some time checking if this feature request has been raised before?

- [X] I checked and didn't find a similar issue

### :package: Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by daniloff200**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

fixes #2316 

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by srujanchikke** [comment]
Hey @daniloff200 ,

Please use the following cargo commands -

`cargo check --all-features`                                                                 
`cargo clippy --all-features` 

ref -[ Cargo commands](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md#cargo-commands), and fix the errors thrown by clippy for the checks to pass .

Please run `cargo +nightly fmt`  to fix failing formatting check

**Comment by daniloff200** [comment]
hi, @srujanchikke . Thanks for your review! I've executed the commands

**Comment by daniloff200** [comment]
hi, @srujanchikke , could you help me with an issue, I have here? I've tried various fixes, but, got a problem with a Some (as here -> https://github.com/juspay/hyperswitch/actions/runs/6459469231/job/17538567259?pr=2492) or with borrowed value.
I'm kinda new to Rust, and tried different approaches, but, it doesn't work

**Comment by srujanchikke** [comment]
> hi, @srujanchikke , could you help me with an issue, I have here? I've tried various fixes, but, got a problem with a Some (as here -> https://github.com/juspay/hyperswitch/actions/runs/6459469231/job/17538567259?pr=2492) or with borrowed value. I'm kinda new to Rust, and tried different approaches, but, it doesn't work

Hi @daniloff200 ,

In Rust, when you move a value, you transfer ownership of that value to another location, and you can no longer use the original location to access the value. 

```
resource_id: types::ResponseId::ConnectorTransactionId(item.response.id.clone()) // Clone the value before moving response.id to another location

File: crates/router/src/connector/shift4/transformers.rs [hunk 1]
```diff
@@ -671,6 +671,8 @@ impl<F>
             types::PaymentsResponseData,
         >,
     ) -> Result<Self, Self::Error> {
+        let token_id = types::ResponseId::ConnectorTransactionId(item.response.token.id.clone());
+        let connector_id = Some(item.response.token.id),;
```
Comment (srujanchikke) [inline crates/router/src/connector/shift4/transformers.rs:675]: Could you remove the additional comma here

File: crates/router/src/connector/shift4/transformers.rs [hunk 2]
```diff
@@ -686,7 +688,7 @@ impl<F>
                 ..item.data.request
             },
             response: Ok(types::PaymentsResponseData::TransactionResponse {
-                resource_id: types::ResponseId::NoResponseId,
+                resource_id: token_id,
```
Comment (srujanchikke) [inline crates/router/src/connector/shift4/transformers.rs:691]: Please revert this change , This is meant to be `NoResponseId`

File: crates/router/src/connector/shift4/transformers.rs [hunk 3]
```diff
@@ -671,6 +671,8 @@ impl<F>
             types::PaymentsResponseData,
         >,
     ) -> Result<Self, Self::Error> {
+        let token_id = types::ResponseId::ConnectorTransactionId(item.response.token.id.clone());
```
Comment (srujanchikke) [inline crates/router/src/connector/shift4/transformers.rs:674]: Shift4ThreeDsResponse don't have field id , Remove token_id

File: crates/router/src/connector/shift4/transformers.rs (hunk 2 above)
Comment (srujanchikke) [inline crates/router/src/connector/shift4/transformers.rs:691]: Revert this change too

**Review by srujanchikke — CHANGES_REQUESTED** [review]
As this PR addressing https://github.com/juspay/hyperswitch/issues/2347 , i am reassigning the issue

**Comment by daniloff200** [comment]
@srujanchikke can you review my pr please? I've managed to fix all issues

File: crates/router/src/connector/shift4/transformers.rs [hunk 4]
```diff
@@ -671,6 +671,7 @@ impl<F>
             types::PaymentsResponseData,
         >,
     ) -> Result<Self, Self::Error> {
+        let token_id = Some(item.response.token.id.clone());
```
Comment (srujanchikke) [inline crates/router/src/connector/shift4/transformers.rs:674]: This token id is not connector response id , it is shift4 3ds token id . revert the changes from shift4 try_from .

**Review by srujanchikke — CHANGES_REQUESTED** [review]
minor changes required , other than that looks good to me !

**Review by srujanchikke — APPROVED** [review]
LGTM

**Review by SanchithHegde — APPROVED** [review]
Looks good to me!

Thanks for the PR, @daniloff200!

**Comment by daniloff200** [comment]
@swangi-kumari  hey!
Thanks for the link, I've submitted the form. One question: can it be the issue, if I live in Ukraine?

**Comment by VedantKhairnar** [comment]
image
Hey @daniloff200 ,
We are using [DHL courier service](https://www.dhl.com/) and they do provide service in Ukraine as shown in the above image. 
If we face any issues, we will reach out to you. If you have any queries, pls let us know.
Thanks.

**Pull request merged**